Sources/App/Models/SemVer.swift

```diff
@@ -5,43 +5,76 @@ struct SemVer: Content, Equatable {
+        self.major = major
+        self.minor = minor
+        self.patch = patch
+        self.preRelease = preRelease
+        self.build = build
+    }
 }
 
 
 extension SemVer: ExpressibleByStringLiteral {
-    // TODO: sas-2020-04-29: This can/should probably be improved...
-    // Or should we just save versions as flat strings?
+    // This initialiser is required for serialisation to the DB - i.e. it needs to have
+    // access to a non-failable conversion from string to SemVer.
+    // SemVer(0, 0, 0) is a safe default.
     init(stringLiteral value: StringLiteralType) {
-        let parts = value.split(separator: ".").map(String.init).compactMap(Int.init)
-        switch parts.count {
-            case 1: self = .init(major: parts[0], minor: 0, patch: 0)
-            case 2: self = .init(major: parts[0], minor: parts[1], patch: 0)
-            case 3: self = .init(major: parts[0], minor: parts[1], patch: parts[2])
-            default: self = .init(major: 0, minor: 0, patch: 0)
-        }
+        self = SemVer(value) ?? .init(0, 0, 0)
     }
 }
 
 
 extension SemVer {
-    // TODO: sas-2020-05-03: This could also be init?(rawValue: String) but that might be confusing
-    // with the existing init(stringLiteral:) that we need to db decoding.
-    static func parse(_ string: String) -> Self? {
-        let parts = string.split(separator: ".").map(String.init)
-        guard parts.allSatisfy({ $0.contained(in: .decimalDigits) }) else { return nil }
-        let numbers = parts.compactMap(Int.init)
-        switch numbers.count {
-            case 1: return .init(major: numbers[0], minor: 0, patch: 0)
-            case 2: return .init(major: numbers[0], minor: numbers[1], patch: 0)
-            case 3: return .init(major: numbers[0], minor: numbers[1], patch: numbers[2])
-            default: return nil
-        }
+    static func isValid(_ string: String) -> Bool { SemVer(string) != nil }
+}
+
+
+extension SemVer: LosslessStringConvertible {
+    init?(_ string: String) {
+        let groups = semVerRegex.matchGroups(string)
+        guard
+            groups.count == semVerRegex.numberOfCaptureGroups,
+            let major = Int(groups[0]),
+            let minor = Int(groups[1]),
+            let patch = Int(groups[2])
+            else { return nil }
+        self = .init(major, minor, patch, groups[3], groups[4])
     }
 
-    static func isValid(_ string: String) -> Bool { parse(string) != nil }
+    var description: String {
+        let pre = preRelease.isEmpty ? "" : "-" + preRelease
+        let bld = build.isEmpty ? "" : "+" + build
+        return "\(major).\(minor).\(patch)\(pre)\(bld)"
+    }
 }
 
 
-extension SemVer: CustomStringConvertible {
-    var description: String { "\(major).\(minor).\(patch)" }
-}
+// Source: https://regex101.com/r/Ly7O1x/3/
+// Linked from https://semver.org
+let semVerRegex = NSRegularExpression(#"""
```

Comment:
Did you say that this regex came from the SPM project itself? If so, let's reference where it came from in the comment.

Reply:
It’s linked from the semver page, at the bottom (i.e. the first link is the link from the semver.org page)